python/lsst/daf/butler/core/dimensions/_config.py

```diff
@@ -0,0 +1,170 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("DimensionConfig",)
+
+from typing import Iterator
+
+from ..config import ConfigSubset
+from .. import ddl
+from .construction import DimensionConstructionBuilder, DimensionConstructionVisitor
+from ._packer import DimensionPackerConstructionVisitor
+from ._skypix import SkyPixConstructionVisitor
+from ._topology import TopologicalSpace
+from .standard import (
+    StandardDimensionElementConstructionVisitor,
+    StandardTopologicalFamilyConstructionVisitor,
+)
+
+
+class DimensionConfig(ConfigSubset):
+    """Configuration that defines a `DimensionUniverse`.
+
+    The configuration tree for dimensions is a (nested) dictionary
+    with five top-level entries:
+
+    - version: an integer version number, used as keys in a singleton registry
+      of all `DimensionUniverse` instances;
+
+    - skypix: a dictionary whose entries each define a `SkyPixSystem`,
+      along with a special "common" key whose value is the name of a skypix
+      dimension that is used to relate all other spatial dimensions in the
+      `Registry` database;
+
+    - elements: a nested dictionary whose entries each define
+      `StandardDimension` or `StandardDimensionCombination`.
+
+    - topology: a nested dictionary with ``spatial`` and ``temporal`` keys,
+      with dictionary values that each define a `StandardTopologicalFamily`.
+
+    - packers: a nested dictionary whose entries define factories for a
+      `DimensionPacker` instance.
+
+    See the documentation for the linked classes above for more information
+    on the configuration syntax.
+    """
+    component = "dimensions"
+    requiredKeys = ("version", "elements", "skypix")
```

Comment:
Will a registry have multiple universe definition in it too? I agree that hash is a pain, could we use user-defined string/tag instead, may be easier to track those than integer versions.

Reply:
No, a Registry instance will only ever have one universe.  User-defined strings means humans like us have to come up with names, and does not sound fun, but you're probably right that it's better than integers, at least in terms of avoiding clashes when there is branching.

Reply:
So if registry only has one universe then multiple universes could only come from different registries? Could we make universe a property of a registry rather than having a singleton dict with all universes?

Reply:
The singleton dict is only needed for pickling, and that pickling is probably only needed for multiprocessing (https://github.com/lsst/daf_butler/blob/master/python/lsst/daf/butler/core/dimensions/universe.py#L333).

So there may be a way to allow multiple versions _or_ pickling but not both (and I think that's probably okay if migration scripts are the only use case for multiple versions), but it'd be a bit tricky.  And we may want the identifiers anyway to organize the migration scripts anyway.

Reply:
That unpickling worries me, it obviously depends on `_instances` being populated by something else. This should work with out current fork-based multiprocesing but that is not the only way to do multiprocessing. Maybe we should stop pickling universe and always get it from registry?

Reply:
The problem is that we need to pickle things like `DataCoordinate` and `DimensionGraph`, which have `universe` attributes.  The `multiprocessing` package does not provide any way to give a registry or universe or anything else to their unpickle logic other than via singletons, as far as I can tell; this is exactly the problem that the [python pickle docs](https://docs.python.org/3/library/pickle.html#persistence-of-external-objects) say to solve with a custom unpickler, but `multiprocessing` does not provide a hook to change the unpickler.

We may be able to slowly remove a lot of those `.universe` attibutes (maybe even all) as `DimensionUniverse` becomes mostly a factory for making `DimensionGraph`, and that might be a big overall improvement.  But there will be a lot of API breakage where we have to add `universe` arguments to many methods instead.